### DISCUSSION TITLE
KeyError: 'path' beim  dwd-mosmix-Abruf

### DISCUSSION THREAD

**Discussion by Xanti19** [roe-dl/weewx-DWD#76] [Q&A] [answered]
Moin,
ich bekomme in weewx mit dem Belchertown-Skin die Wettervorhersagen nicht geladen. Der Aufruf mit --verbose zeigt dies:

**************************************************************************
root@weewxTest:/usr/local/bin# ./dwd-mosmix --weewx --daily --hourly 10403 --verbose
Using configuration file /etc/weewx/weewx.conf
Traceback (most recent call last):
  File "/usr/local/bin/./dwd-mosmix", line 1959, in <module>
    dwd = DwdMosmix(config,options.verbose)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/bin/./dwd-mosmix", line 351, in __init__
    self.target_path = config_dict['DeutscherWetterdienst']['path']
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
  File "/usr/lib/python3/dist-packages/configobj/__init__.py", line 554, in __getitem__
    val = dict.__getitem__(self, key)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'path'
*******************************************************************************

Ist mit KeyError: 'path' jetzt die Angabe in der weewx.conf bei [WeatherSerices] gemeint? Hier habe ich aktuell dies eingetragen:

*******************************************************************************
[DeutscherWetterdienst]
    # Konfiguration für dwd-cap-warnings
    [[warning]]
        icons='../dwd/warn_icons_50x50'
        states= Nordrhein-Westfalen
        resolution = county|city
        [[[counties]]]
              Stadt Mönchengladbach = MG
        [[[cities]]]
              Stadt Mönchengladbach = MG
[WeatherServices]
    # Verzeichnis, in das die Dateien gespeichert werden sollen
    path='/etc/weewx/skins/Belchertown/dwd'
    # Warnungen (dwd-cap-warnings, bbk-warnings, msc-warnings)

   [[forecast]]
        # Speicherort der Symbole auf dem Web-Server
        icons='../images'
        # zu verwendender Symbolsatz: belchertown, dwd, aeris
        icon_set = dwd
        # für welche Orientierungen soll die HTML-Datei erzeugt werden?
        orientation = h,v
        # Sollen die Meßgrößensymbole in die Tabelle aufgenommen werden?
        #show_obs_symbols = True # optional
        # Soll die Meßgrößenbeschreibung in die Tabelle aufgenommen werden?
        #show_obs_description = False # optional
        # Soll der Ortsname über die Tabelle geschrieben werden?
        #show_placemark = True # optional
        # Beispiele
        [[[10403]]]
            # Wettervorhersage
            provider = DWD
            model = MOSMIX
	    
            # optional darzustellende Wetterelemente anstelle des defaults
            #observations_daily = ww, TTTmax, TTTmin, windchill, FFavg, DDavg, RR1c, Rd10, Neffavg, RSunD, Rad1hsum, VVmin
            #observations_hourly = TTT, humidity , FF , DD , RR1c , R101 , PPPP , Rad1h
            # Was würde eine PV-Anlage an dem Tag bringen? 
            #pv_factor = 10.0 # Fläche * Wirkungsgrad 


   [[warning]]
        #icons = ... # Optional, für alle Provider
        #bbk_icons = ... # Optional, nur für BBK
        #bbk_logos = ... # Optional, nur für BBK
        #bbk_include_dwd ... # Optional, nur für BBK
        #dwd_icons = ... # Optional, nur für DWD
        #dwd_status_url = ... # Optional, nur für DWD
        #dwd_diff_url = ... # Optional, nur für DWD
        # Beispiele

     # Konfiguration für die --belchertown Option von dwd-mosmix
     [[Belchertown]]
        # Name des Abschnittes der Belchertown-Skin in [StdReport]
        section = Belchertown
        # Warndatei aus dem Abschnitt [[warnings]]
        warnings = DL
        # Vorhersagedatei aus dem Aufruf 'dwd-mosmix'
        forecast = 10403
        # schließe Warnungen ein, die die angegebene Zahl von Sekunden in
        # der Zukunft gültig werden (optional)
        #include_advance_warnings = 0 
        # Quelle für Luftqualtitätsdaten (optional)
        # mögliche Werte: aeris, ubaXXXX
        #aqi_source = ...
        # Sprache für die Himmelsrichtungen (optional)
        # mögliche Werte: de, en, fr, it, cz, es, nl, no, gr
        #compass_lang = 'de' # optional
    [[download]]
        # allgemeines Download-Interface zum Herunterladen von Karten und
        # Vorhersagen, die die Wetterdienste bereitstellen
        [[[Download3]]]
            # Warnkarte mit straßenschildähnlichen Symbolen
            provider = DWD
            model = warning-map-with-symbols
            area = EM
        [[[Download4]]]
            # Warnkarte nur mit Farben
            provider = DWD
            model = warning-map
            area = nrw
**********************************************************************************

Die weewx-Installation befindet sich auf einem Proxmox in einem LXC-Container. Müssen deswegen vielleicht die Pfadangaben anders lauten?
Der Abruf der dwd-cap-warnings funktioniert.

**Comment by roe-dl** [comment]
Ich würde die Zeile `path='/etc/weewx/skins/Belchertown/dwd'` noch einmal unter `[DeutscherWetterdienst]` kopieren.

**Comment by Xanti19** [comment]
Danke für die schnelle Rückmeldung!
Das hat schon mal geholfen. Musste auch den Abschnitt [[forecast]] unter [DeutscherWetterdienst] positionieren.  Im DWD-Verzeichnis wurden nun die Dateien forecast-10403.json , forecast-10403.inc , und forecast-10403-hourly.inc erstellt. 
Den Abschnitt [[Belchertown]] habe ich auch unter [DeutscherWetterdienst] verschoben. Hier gibt verbose noch einen Fehler aus:  AttributeError: 'DwdMosmix' object has no attribute 'belchertown_compasslang' , auch wenn ich in der weewx.conf compass_lang='de' aktiviere. Eine forecast.json wird nicht erstellt.

**Comment by roe-dl** [comment]
Bei mir ist der Abschnitt `[[forecast]]` unter `[WeatherServices]`. 

Den Fehler mit `belchertown_compasslang` kann ich nicht so recht einordnen. Vielleicht erschließt es sich mir, wenn ich die komplette Ausgabe des Programms unter `--verbose` sehe.

Achtung! Die `forecast.json`-Datei für Belchertown wird an einer anderen Stelle gespeichert als die übrigen Dateien, nämlich dort, wo die Belchertown-Skin sie erwartet, unter `json`.

**Comment by Xanti19** [comment]
Die komplette Ausgabe sieht so aus:

root@weewxTest:/usr/local/bin# ./dwd-mosmix --weewx --belchertown 10403 --verbose
Using configuration file /etc/weewx/weewx.conf
-- configuration data ----------------------------------
log success:       True
log failure:       True
target path:       /etc/weewx/skins/Belchertown/dwd
horiz. tab:        True
vertical tab:      True
icon set:          4
station location:  lat 51.202 lon 6.342 alt 74.0
aeris api id:      None
aeris api secret:  None
--------------------------------------------------------
successfully downloaded https://opendata.dwd.de/weather/local_forecasts/mos/MOSMIX_L/single_stations/10403/kml/MOSMIX_L_LATEST_10403.kmz
-- MOSMIX_L_2024091415_10403.kml --
processing KML file
KML file processed, 1 placemarks found
what to output: ['belchertown']
start loop over placemarks
process placemark id "10403" name "MOENCHENGLADBACH"
calculate daily forecast
calculate 3hr forecast
<ephem.Observer date='1970/1/1 00:00:00' epoch='1970/1/1 00:00:00' lon='6:30:00.0' lat='51:13:48.0' elevation=37.0m horizon=0:00:00.0 temp=15.0C pressure=1010.0mBar>
belchertown
geo file '/etc/weewx/skins/Belchertown/dwd/geo-10403.json' succuessfully loaded
now 20:07:15 timestep 21:00:00
<ephem.Observer date='2024/9/14 19:00:00' epoch='2024/9/14 19:00:00' lon='6:30:00.0' lat='51:13:48.0' elevation=37.0m horizon=0:00:00.0 temp=15.0C pressure=1010.0mBar>
Traceback (most recent call last):
  File "/usr/local/bin/./dwd-mosmix", line 1983, in <module>
    dwd.forecast_all(mmos,output,options.dry_run,lang=options.lang)
  File "/usr/local/bin/./dwd-mosmix", line 1819, in forecast_all
    self.forecast_placemark(placemark, mos['ForecastTimeSteps'], issue, output, dryrun, lang)
  File "/usr/local/bin/./dwd-mosmix", line 1800, in forecast_placemark
    self.belchertown(placemark,days,recs3hr,timesteps,issue,dryrun)
  File "/usr/local/bin/./dwd-mosmix", line 1155, in belchertown
    'windDir':compass(nround(placemark['Forecast']['DD'][idx]),self.belchertown_compasslang,False),
                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'DwdMosmix' object has no attribute 'belchertown_compasslang'
root@weewxTest:/usr/local/bin#

Die Zeile beginnend mit <ephem.Observer date=' wird 80 mal ausgegeben,

**Comment by roe-dl** [comment]
> Die Zeile beginnend mit <ephem.Observer date=' wird 80 mal ausgegeben,

Das ist korrekt, daß das so oft ist. Das muß für jede Stunde berechnet werden. Was mir weniger gefällt, ist, daß bei `date` dort kein aktuelles Datum angezeigt wird.

> `File "/usr/local/bin/./dwd-mosmix", line 1155, in belchertown`
> `'windDir':compass(nround(placemark['Forecast']['DD'][idx]),self.belchertown_compasslang,False),`

In Zeile 1155 finde ich in der aktuellen Version nicht die Zeile, die darunter angegeben ist. Ist das wirklich die aktuelle Version des Programms?

> `dwd-mosmix --weewx --belchertown 10403 --verbose`

Bei mir läuft der Aufruf fehlerfrei durch. Dabei ist die Ausgabe wesentlich länger.

**Comment by Xanti19** [comment, marked as answer]
Oh nein, hatte tatsächlich noch eine alte Version beim Kopieren erwischt. Jetzt läuft es durch.:

INFO geo file '/etc/weewx/skins/Belchertown/dwd/geo-10403.json' succuessfully loaded
INFO now 22:34:51 timestep 23:00:00
INFO write Belchertown JSON file to /var/www/html/weewx/belchertown/json/forecast.json
INFO placemark id "10403" processed
INFO end loop over placemarks

Ist es richtig, dass die forecast.json direkt nach /var.... geschrieben wird? Wie binde ich die Informationen jetzt bei Belchertown ein?
Sorry und danke für den späten Support.

**Comment by Xanti19** [comment]
Habe es gerade gefunden. In der weewx.conf musste ich noch forecast_enabled = 1 einstellen. Jetzt stehen auf der Startseite 3 Intervalle 1,3 und 24 zur Verfügung. ;-)

**Comment by roe-dl** [comment]
Das ist schön. Dann ist es ja geschafft.
